### PR TITLE
[ZCash] Fetch consensus branch id from the backend

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/42951

Resolves

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/zcash/zcash_transaction_complete_manager.cc [hunk 1]
```diff
@@ -54,9 +54,9 @@ void ZCashTransactionCompleteManager::CompleteTransaction(
     const ZCashTransaction& transaction,
     const mojom::AccountIdPtr& account_id,
     CompleteTransactionCallback callback) {
-  zcash_wallet_service_->zcash_rpc().GetLatestBlock(
+  zcash_wallet_service_->zcash_rpc().GetLightdInfo(
       chain_id,
-      base::BindOnce(&ZCashTransactionCompleteManager::OnGetLatestBlockHeight,
+      base::BindOnce(&ZCashTransactionCompleteManager::OnGetLightdInfo,
                      weak_ptr_factory_.GetWeakPtr(),
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_transaction_complete_manager.cc:60]: How frequently branch id changes? Does it make sense to cache it per browser instance, or persist it per network?

File: components/brave_wallet/browser/zcash/zcash_wallet_service.cc [hunk 2]
```diff
@@ -334,7 +334,6 @@ void ZCashWalletService::CompleteTransactionDone(
                             result.error());
     return;
   }
-
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_wallet_service.cc:337]: need this change?

File: components/brave_wallet/browser/zcash/zcash_serializer.cc [hunk 3]
```diff
@@ -93,6 +91,20 @@ std::array<uint8_t, 32> HashScriptPubKeys(const ZCashTransaction& tx) {
...
+      .subspan(0, sizeof(kTxHashPersonalizerPrefix) - 1)
+      .copy_from(base::byte_span_from_cstring(kTxHashPersonalizerPrefix));
+  base::span(result)
+      .subspan(sizeof(kTxHashPersonalizerPrefix) - 1, sizeof(consensusBranchId))
+      .copy_from(base::byte_span_from_ref(base::numerics::U32FromLittleEndian(
+          base::byte_span_from_ref(consensusBranchId))));
+  return result;
+}
+
 }  // namespace
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_serializer.cc:109]: That should be done with SpanWriter
something like
```
auto span_writer = base::SpanWriter(base::span(result));
span_writer.Write(base::byte_span_from_cstring(kTxHashPersonalizerPrefix));
span_writer.WriteU32LittleEndian(tx.consensus_brach_id())
```

File: components/brave_wallet/browser/zcash/zcash_serializer.cc [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include <utility>
 
 #include "base/containers/span.h"
+#include "base/containers/to_vector.h"
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_serializer.cc:13]: ~#include "base/containers/to_vector.h"~

File: components/brave_wallet/browser/zcash/zcash_transaction_complete_manager.cc (hunk 1 above)
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_transaction_complete_manager.cc:60]: Several times a year i guess, we have to check whether it has been changed anyway

**Review by supermassive — APPROVED** [review]
wallet core lgtm

**Comment by srirambv** [comment]
Verification passed on

Brave | 1.75.145 Chromium: 132.0.6834.57 (Official Build) nightly (64-bit)
-- | --
Revision | 4ecdecd1626892f9f317e9d8154c50cf8459d3bf
OS | Windows 11 Version 23H2 (Build 22631.4602)

-  Verified steps from issue description
- Verified sending ZCash transaction from transparent account to a shielded account works as expected

https://github.com/user-attachments/assets/6075b3f5-5d0f-4640-b12d-24f7738f097e

**Pull request merged**